### PR TITLE
Fix: Support local cache of APK index

### PR DESCRIPTION
**Pull request by rahul2393**
[Issue](https://github.com/aquasecurity/fanal/issues/124)
* Added support of reading from local file for air-gapped environments
* Added env to skip APK index parsing

### CODE REVIEW

**Comment by rahul2393** [comment]
@knqyf263 Please help review

File: analyzer/command/apk/apk_test.go [hunk 1]
```diff
@@ -162,14 +162,7 @@ func TestAnalyze(t *testing.T) {
...
 	for testName, v := range tests {
-		f, err := os.Open(v.apkIndexArchivePath)
-		if err != nil {
-			t.Fatalf("unexpected error: %s", err)
-		}
-		apkIndexArchive := &apkIndex{}
-		if err = json.NewDecoder(f).Decode(apkIndexArchive); err != nil {
-			t.Fatalf("unexpected error: %s", err)
-		}
+		apkIndexArchiveURL = v.apkIndexArchivePath
```
Comment (knqyf263) [inline analyzer/command/apk/apk_test.go:165]: Could you add a test using [httptest](https://golang.org/pkg/net/http/httptest/)? I'd like to test `https://` and `file://` both.

File: analyzer/command/apk/apk.go [hunk 2]
```diff
@@ -79,17 +86,25 @@ func (a alpineCmdAnalyzer) fetchApkIndexArchive(targetOS types.OS) (*apkIndex, e
 	}
 
 	url := fmt.Sprintf(apkIndexArchiveURL, osVer)
-	resp, err := http.Get(url)
-	if err != nil {
-		return nil, xerrors.Errorf("failed to fetch APKINDEX archive: %w", err)
+	var reader io.Reader
+	if strings.HasPrefix(url, "file://") {
+		content, err := ioutil.ReadFile(strings.Replace(url, "file://", "", 1))
```
Comment (knqyf263) [inline analyzer/command/apk/apk.go:91]:
```suggestion
		content, err := os.Open(strings.TrimPrefix(url, "file://"))
```

File: analyzer/command/apk/apk.go [hunk 3]
```diff
@@ -79,17 +86,25 @@ func (a alpineCmdAnalyzer) fetchApkIndexArchive(targetOS types.OS) (*apkIndex, e
...
+		if err != nil {
+			return nil, xerrors.Errorf("failed to fetch APKINDEX archive: %w", err)
```
Comment (knqyf263) [inline analyzer/command/apk/apk.go:93]:
```suggestion
			return nil, xerrors.Errorf("failed to read APKINDEX archive file: %w", err)
```

File: analyzer/command/apk/apk.go [hunk 4]
```diff
@@ -79,17 +86,25 @@ func (a alpineCmdAnalyzer) fetchApkIndexArchive(targetOS types.OS) (*apkIndex, e
...
+		}
+		reader = bytes.NewReader(content)
```
Comment (knqyf263) [inline analyzer/command/apk/apk.go:95]: unnecessary

File: analyzer/analyzer_test.go [hunk 5]
```diff
@@ -349,9 +354,11 @@ func TestAnalyzeConfig(t *testing.T) {
...
 	}
 	tests := []struct {
-		name string
-		args args
-		want []types.Package
+		name     string
+		args     args
+		want     []types.Package
+		setEnv   func()
+		unsetEnv func()
```
Comment (knqyf263) [inline analyzer/analyzer_test.go:361]: Where is this used?

File: analyzer/command/apk/apk.go [hunk 6]
```diff
@@ -257,5 +272,9 @@ func (a alpineCmdAnalyzer) guessVersion(apkIndexArchive *apkIndex, originalPkgs
 }
 
 func (a alpineCmdAnalyzer) Required(targetOS types.OS) bool {
+	if builtinos.Getenv(skipAPKIndex) != "" {
```
Comment (knqyf263) [inline analyzer/command/apk/apk.go:275]: Can we disable this analyzer by default and enable it when a CLI option is passed? It is best if we can do it with the following option.
https://github.com/aquasecurity/trivy/blob/96af6dc49943731cfa398d1e9a4fa4d566566e84/internal/app.go#L126-L130

File: analyzer/analyzer.go [hunk 7]
```diff
@@ -63,7 +63,7 @@ func (r AnalyzeReturn) ConvertToResult(analyzerType, filePath string) *AnalysisR
 type analyzer interface {
 	Name() string
 	Analyze(content []byte) (AnalyzeReturn, error)
-	Required(filePath string, info os.FileInfo) bool
+	Required(filePath string, info os.FileInfo, inspectRemovedPackages bool) bool
```
Comment (knqyf263) [inline analyzer/analyzer.go:66]: I don't think we have to add `inspectRemovedPackages` to all analyzers. At the moment, let's keep `file://` support in this PR and revert other changes. After merging this PR, we can take time to discuss `inspectRemovedPackages` in the dedicated PR. I'm sorry, but could you leave `file://` support only once?

**Comment by rahul2393** [comment]
@knqyf263 Done

File: analyzer/command/apk/apk_test.go [hunk 8]
```diff
@@ -162,21 +307,17 @@ func TestAnalyze(t *testing.T) {
...
-		actual, _ := analyzer.Analyze(v.args.targetOS, v.args.configBlob)
-		sort.Slice(actual, func(i, j int) bool {
-			return actual[i].Name < actual[j].Name
+		t.Run(testName, func(t *testing.T) {
+			apkIndexArchiveURL = v.apkIndexArchivePath
+			actual, _ := analyzer.Analyze(v.args.targetOS, v.args.configBlob)
+			sort.Slice(actual, func(i, j int) bool {
+				return actual[i].Name < actual[j].Name
+			})
+			if !reflect.DeepEqual(v.expected, actual) {
```
Comment (knqyf263) [inline analyzer/command/apk/apk_test.go:316]: This is an old test. We recently use `testify/assert`. Let's replace it this time.

```suggestion
			assert.Equal(t, v.expected, actual)
```

**Comment by rahul2393** [comment]
@knqyf263 Done

File: analyzer/command/apk/apk_test.go [hunk 9]
```diff
@@ -2,18 +2,35 @@ package apk
...
 	"github.com/kylelemons/godebug/pretty"
 
 	"github.com/aquasecurity/fanal/types"
 )
 
 func TestAnalyze(t *testing.T) {
+	testServer := httptest.NewServer(http.HandlerFunc(func(res http.ResponseWriter, req *http.Request) {
+		content, err := ioutil.ReadFile("testdata/history_v3.9.json")
+		if err != nil {
+			res.WriteHeader(http.StatusInternalServerError)
```
Comment (knqyf263) [inline analyzer/command/apk/apk_test.go:25]: unnecessary

Comment (rahul2393) [reply analyzer/command/apk/apk_test.go]: @knqyf263 Thanks, removed

**Review by knqyf263 — APPROVED** [review]

**Comment by patrick-stephens** [comment]
Thanks for resolving this, I'll aim to deploy as soon as I can.

**Pull request merged**